Add clarification note for flash hashes

albertoalmagro opened the pull request:
### Summary

When saving hash values in the flash, depending on session storage settings the result after serialization and deserialization can be different as expected. To avoid the confusion that this may cause I have added a note to the documentation.

See issue #33447

schneems commented:
I think it’s a fine idea to add this as a caveat. I’m not sure this is the best place though. Is there anywhere in the docs where we show storing alternative objects? Maybe in a method doc somewhere? We could ask the dev in the other ticket where they got the idea.

The other thing I’m not sure about is linking to an issue in the docs. I think it would be better if we can summarize the issue in a sentence or two. Clicking through and reading a raw issue is a bit of a dense experience.

albertoalmagro commented:
Thanks @schneems for your feedback! I like it ❤️Following your suggestions I have just asked him about that.

kaspth commented:
> The other thing I’m not sure about is linking to an issue in the docs. I think it would be better if we can summarize the issue in a sentence or two. Clicking through and reading a raw issue is a bit of a dense experience.

👍, guides are meant to be self-contained (at least when it comes to our bug tracker), so we shouldn't link to GitHub issues within them.

albertoalmagro commented:
@schneems @kaspth You are right, I have changed it. I took the idea from the [Upgrading Ruby on Rails guide](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#rails-controller-testing), but while linking issues there is justified because it refers to particular changes and people may want to know deeper about them, I fully agree with you that in this case having a short explanation in the guide is much better. Thanks for your guidance ❤️

CHTJonas commented:
Hi guys! Not quite sure where or why I had the idea to store a hash in the flash rather than just, say, a string. I didn't actually realise that Rails would serialise it and therefore the symbol keys would get stringified. But now it's been pointed out to me it makes perfect sense!

In terms of the docs, maybe a small separate section would be useful. As a learning exercise I actually created a custom object to store in the flash just to see what would happen:
```ruby
module MyWebApp
  class Alert
    attr_accessor :class, :message

    def initialize
      yield(self) if !block_given?
    end

  end
end
```
Whilst storing weird objects is perhaps a bit more of an esoteric use of the flash, maybe it does warrant an extra section in the docs? I'll leave that up to you guys to decide! Thanks for all the help and support 😄❤️

albertoalmagro commented:
@schneems I found it. In the Flash module's documentation we say that a Hash is an object that could be stored in it. Please follow this link to see [ActionDispatch::Flash ](https://github.com/rails/rails/blob/master/actionpack/lib/action_dispatch/middleware/flash.rb#L6)'s documentation.

Now that we know that it was intentional, I wouldn't warn about using a non string object, but stick to the warning note. I think it makes sense to move the warning note together with the `Flash` module's documentation. What do you think?

CHTJonas commented:
Yeah I think a warning if you try and store an object in the flash other than one of String, Array or Hash wouldn't go amiss 👍

schneems commented:
I like that, also add an escape valve to say “I know the risks and want to
do it anyway”.
On Mon, Jul 30, 2018 at 4:05 PM Charlie Jonas <notifications@github.com>
wrote:

> Yeah I think a warning if you try and store an object in the flash other
> than one of String, Array or Hash wouldn't go amiss 👍
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/33465#issuecomment-408992113>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AADpYKVg75puYmoLvwVnuK6lCtHf8fyNks5uL2cJgaJpZM4VlNiR>
> .
>

schneems commented:
This looks great! 

What are you thinking for an escape valve? Maybe a wrapper class such as FlashCustomValue that holds another class value?

schneems commented on actionpack/lib/action_dispatch/middleware/flash.rb:
```diff
@@ -284,11 +287,32 @@ def now_is_loaded?
         end
+
+        def supported_primitive_type?(v)
+          v.nil? || SUPPORTED_PRIMITIVE_TYPES.any? { |type| v.is_a?(type) }
```
Please switch from iteration over an array and using any? To using a hash with the class names as keys. It will be faster.

albertoalmagro replied:
@schneems I have benchmarked this and I don't see a big benefit in this case, maybe because the array has only 3 positions. If there isn't a big different I would prefer to stick to the Array version as it seems clearer to me.

Here you have the benchmark:
```ruby
require "Benchmark"
include Benchmark

SUPPORTED_PRIMITIVE_ARRAY = [String, Array, Hash].freeze
SUPPORTED_PRIMITIVE_HASH  = { String => nil, Array => nil, Hash => nil }

["String", [1, 2, 3], { a: 1, b: 2}].each do |value|
  puts "\n****************************"
  puts "Benchmarking value: #{value}"
  puts

  bmbm(6) do |test|
    test.report("Array: ") do
      SUPPORTED_PRIMITIVE_ARRAY.any? { |type| value.is_a?(type) }
    end

    test.report("Hash: ") do
      SUPPORTED_PRIMITIVE_HASH.any? { |k, _| value.is_a?(k) }
    end

    test.report("OR: ") do
      value.is_a?(String) || value.is_a?(Array) || value.is_a?(Hash)
    end
  end
end
```
And the results:
```bash
2.5.1 ~/ $ ruby bench_array_hash_or.rb

****************************
Benchmarking value: String

Rehearsal -------------------------------------------
Array:    0.000010   0.000006   0.000016 (  0.000006)
Hash:     0.000007   0.000002   0.000009 (  0.000009)
OR:       0.000003   0.000001   0.000004 (  0.000004)
---------------------------------- total: 0.000029sec

              user     system      total        real
Array:    0.000008   0.000002   0.000010 (  0.000005)
Hash:     0.000007   0.000002   0.000009 (  0.000006)
OR:       0.000005   0.000002   0.000007 (  0.000023)

****************************
Benchmarking value: [1, 2, 3]

Rehearsal -------------------------------------------
Array:    0.000005   0.000001   0.000006 (  0.000004)
Hash:     0.000007   0.000004   0.000011 (  0.000018)
OR:       0.000003   0.000001   0.000004 (  0.000003)
---------------------------------- total: 0.000021sec

              user     system      total        real
Array:    0.000024   0.000002   0.000026 (  0.000006)
Hash:     0.000008   0.000001   0.000009 (  0.000005)
OR:       0.000005   0.000001   0.000006 (  0.000005)

****************************
Benchmarking value: {:a=>1, :b=>2}

Rehearsal -------------------------------------------
Array:    0.000004   0.000001   0.000005 (  0.000005)
Hash:     0.000004   0.000001   0.000005 (  0.000004)
OR:       0.000004   0.000001   0.000005 (  0.000005)
---------------------------------- total: 0.000015sec

              user     system      total        real
Array:    0.000004   0.000001   0.000005 (  0.000004)
Hash:     0.000004   0.000001   0.000005 (  0.000004)
OR:       0.000004   0.000001   0.000005 (  0.000003)
```

schneems replied:
You’re still iterating through the hash by calling any? ans checking each key. Instead use it as a hash:

    SUPPORTED_PRIMITIVE_HASH[value.class]


For the detection to work you would also need the values to be truthy.

I recommend using the benchmark-ips gem over Benchmark.bmbm.

albertoalmagro replied:
OK, I got it wrong, this should give us a hint, now using benchmark-ips gem ❤️ 
```ruby
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "benchmark-ips"
end

require "benchmark/ips"

SUPPORTED_PRIMITIVE_ARRAY = [String, Array, Hash].freeze
SUPPORTED_PRIMITIVE_HASH  = { String => true, Array => true, Hash => true }

["String", [1, 2, 3], { a: 1, b: 2}].each do |value|
  puts "\n****************************"
  puts "Benchmarking value: #{value}"
  puts

  Benchmark.ips do |test|
    test.report("Array: ") do
      SUPPORTED_PRIMITIVE_ARRAY.any? { |type| value.is_a?(type) }
    end

    test.report("Hash: ") do
      SUPPORTED_PRIMITIVE_HASH[value.class]
    end

    test.report("OR: ") do
      value.is_a?(String) || value.is_a?(Array) || value.is_a?(Hash)
    end

    test.compare!
  end
end
```
And the results:
```bash
2.5.1 ~ $ ruby bench_array_hash_or.rb
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Using benchmark-ips 2.7.2
Using bundler 1.16.1

****************************
Benchmarking value: String

Warming up --------------------------------------
             Array:    272.789k i/100ms
              Hash:    192.142k i/100ms
                OR:    336.287k i/100ms
Calculating -------------------------------------
             Array:       6.625M (± 4.6%) i/s -     33.280M in   5.035105s
              Hash:       3.442M (± 4.2%) i/s -     17.293M in   5.033748s
                OR:      10.453M (± 7.6%) i/s -     52.124M in   5.024647s

Comparison:
                OR: : 10453063.7 i/s
             Array: :  6624951.4 i/s - 1.58x  slower
              Hash: :  3441698.7 i/s - 3.04x  slower


****************************
Benchmarking value: [1, 2, 3]

Warming up --------------------------------------
             Array:    214.251k i/100ms
              Hash:    183.673k i/100ms
                OR:    295.872k i/100ms
Calculating -------------------------------------
             Array:       4.921M (± 4.7%) i/s -     24.639M in   5.018891s
              Hash:       3.444M (± 3.8%) i/s -     17.265M in   5.021349s
                OR:       8.834M (± 3.8%) i/s -     44.381M in   5.031975s

Comparison:
                OR: :  8833775.5 i/s
             Array: :  4920635.0 i/s - 1.80x  slower
              Hash: :  3443628.5 i/s - 2.57x  slower


****************************
Benchmarking value: {:a=>1, :b=>2}

Warming up --------------------------------------
             Array:    205.368k i/100ms
              Hash:    193.242k i/100ms
                OR:    287.559k i/100ms
Calculating -------------------------------------
             Array:       3.712M (± 4.3%) i/s -     18.688M in   5.044509s
              Hash:       3.443M (± 4.8%) i/s -     17.199M in   5.008349s
                OR:       7.544M (± 4.0%) i/s -     37.670M in   5.001988s

Comparison:
                OR: :  7544189.8 i/s
             Array: :  3711993.7 i/s - 2.03x  slower
              Hash: :  3442707.2 i/s - 2.19x  slower
```
After launching it several times I always get similar results. As there are only 3 supported classes, I think I will go for the OR which always performs better.

schneems replied:
My responses went to the wrong place here are the results with compare by identity

```
Benchmarking value: String

Warming up --------------------------------------
             Array:    268.224k i/100ms
              Hash:    318.214k i/100ms
                OR:    335.285k i/100ms
Calculating -------------------------------------
             Array:       6.201M (± 2.3%) i/s -     31.114M in   5.020387s
              Hash:       9.245M (± 3.6%) i/s -     46.459M in   5.032277s
                OR:       9.342M (± 5.6%) i/s -     46.605M in   5.004818s

Comparison:
                OR: :  9342494.7 i/s
              Hash: :  9245341.0 i/s - same-ish: difference falls within error
             Array: :  6200932.3 i/s - 1.51x  slower


****************************
Benchmarking value: [1, 2, 3]

Warming up --------------------------------------
             Array:    231.245k i/100ms
              Hash:    316.924k i/100ms
                OR:    302.787k i/100ms
Calculating -------------------------------------
             Array:       4.549M (± 2.6%) i/s -     22.893M in   5.036167s
              Hash:       9.033M (± 4.1%) i/s -     45.320M in   5.026575s
                OR:       7.770M (± 3.1%) i/s -     39.060M in   5.031966s

Comparison:
              Hash: :  9033053.8 i/s
                OR: :  7770054.8 i/s - 1.16x  slower
             Array: :  4548977.5 i/s - 1.99x  slower


****************************
Benchmarking value: {:a=>1, :b=>2}

Warming up --------------------------------------
             Array:    198.171k i/100ms
              Hash:    318.380k i/100ms
                OR:    284.343k i/100ms
Calculating -------------------------------------
             Array:       3.425M (± 2.5%) i/s -     17.241M in   5.036619s
              Hash:       9.075M (± 4.5%) i/s -     45.528M in   5.029719s
                OR:       6.503M (± 3.1%) i/s -     32.699M in   5.033228s

Comparison:
              Hash: :  9075060.8 i/s
                OR: :  6503066.2 i/s - 1.40x  slower
             Array: :  3425284.8 i/s - 2.65x  slower
```

The  only difference is adding this line:

```
SUPPORTED_PRIMITIVE_HASH.compare_by_identity.rehash
```

Hash is either the fastest or equal to or in all the cases. This will only improve if we add another supported class. Which, if we're going to support floats and numeric types I think we will.

rafaelfranca commented:
Nobody is going to see that warning in the middle of the logs. We we want to protect people we should deprecate and later raise an exception.

albertoalmagro commented:
@schneems Sure, it's just that I don't understand the idea of the escape valve 100%. Could you please tell me more or less what are you thinking about so I can get it better?

@rafaelfranca I like the idea! I can raise a deprecation warning in case the type isn't the expected.

albertoalmagro committed:
[ci skip] Add note for flash Hash values

When storing hash values in the flash, depending on session storage
settings the result after serialization and deserialization can be
different as expected. See issue #33447

albertoalmagro commented:
@schneems @rafaelfranca I have updated the PR to raise a deprecation warning instead of logging.

schneems commented:
Sorry to get jargon-y on you. By “escape valve” I meant add a way for a user to say to us “I know this is a bad idea but I want to do it anyway. For example look at the bottom section of this changelog https://devcenter.heroku.com/changelog-items/1437.

So I’m suggesting we add a class that people can wrap their custom objects in. When the check runs against it we will see that it’s an approved class and not emit an error or warning.

About the deprecation: we didn’t do a great job of letting you know what we wanted. In this case rails master is pointing at version 6.0. What we want is for the behavior in master to raise an error.

Then in a separate PR we can add a deprecation notice and target the 5.2 branch. So that anyone upgrading will first get the deprecation notice before they upgrade and start getting errors.

I broke the screen on my laptop so sorry for the delayed responses. Let me know if you have any more questions. I can get back to a real computer likely on Monday.

schneems commented:
Call rehash on the hash before running the tests also try adding a 4th key
since we plan on adding an “escape valve” class.


On Sun, Aug 5, 2018 at 7:51 AM Alberto Almagro Sotelo <
notifications@github.com> wrote:

> *@albertoalmagro* commented on this pull request.
> ------------------------------
>
> In actionpack/lib/action_dispatch/middleware/flash.rb
> <https://github.com/rails/rails/pull/33465#discussion_r207735370>:
>
> >          def stringify_array(array) # :doc:
>            array.map do |item|
>              item.kind_of?(Symbol) ? item.to_s : item
>            end
>          end
> +
> +        def supported_primitive_type?(v)
> +          v.nil? || SUPPORTED_PRIMITIVE_TYPES.any? { |type| v.is_a?(type) }
>
> OK, I got it wrong, this should give us a hint, now using benchmark-ips
> gem ❤️
>
> begin
>   require "bundler/inline"rescue LoadError => e
>   $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
>   raise eend
>
> gemfile(true) do
>   source "https://rubygems.org"
>
>   git_source(:github) { |repo| "https://github.com/#{repo}.git" }
>
>   gem "benchmark-ips"end
> require "benchmark/ips"
> SUPPORTED_PRIMITIVE_ARRAY = [String, Array, Hash].freezeSUPPORTED_PRIMITIVE_HASH  = { String => true, Array => true, Hash => true }
>
> ["String", [1, 2, 3], { a: 1, b: 2}].each do |value|
>   puts "\n****************************"
>   puts "Benchmarking value: #{value}"
>   puts
>
>   Benchmark.ips do |test|
>     test.report("Array: ") do
>       SUPPORTED_PRIMITIVE_ARRAY.any? { |type| value.is_a?(type) }
>     end
>
>     test.report("Hash: ") do
>       SUPPORTED_PRIMITIVE_HASH[value.class]
>     end
>
>     test.report("OR: ") do
>       value.is_a?(String) || value.is_a?(Array) || value.is_a?(Hash)
>     end
>
>     test.compare!
>   endend
>
> And the results:
>
> 2.5.1 ~ $ ruby bench_array_hash_or.rb
> Fetching gem metadata from https://rubygems.org/.
> Resolving dependencies...
> Using benchmark-ips 2.7.2
> Using bundler 1.16.1
> ****************************
> Benchmarking value: String
>
> Warming up --------------------------------------
>              Array:    272.789k i/100ms
>               Hash:    192.142k i/100ms
>                 OR:    336.287k i/100ms
> Calculating -------------------------------------
>              Array:       6.625M (± 4.6%) i/s -     33.280M in   5.035105s
>               Hash:       3.442M (± 4.2%) i/s -     17.293M in   5.033748s
>                 OR:      10.453M (± 7.6%) i/s -     52.124M in   5.024647s
>
> Comparison:
>                 OR: : 10453063.7 i/s
>              Array: :  6624951.4 i/s - 1.58x  slower
>               Hash: :  3441698.7 i/s - 3.04x  slower
>
> ****************************
> Benchmarking value: [1, 2, 3]
>
> Warming up --------------------------------------
>              Array:    214.251k i/100ms
>               Hash:    183.673k i/100ms
>                 OR:    295.872k i/100ms
> Calculating -------------------------------------
>              Array:       4.921M (± 4.7%) i/s -     24.639M in   5.018891s
>               Hash:       3.444M (± 3.8%) i/s -     17.265M in   5.021349s
>                 OR:       8.834M (± 3.8%) i/s -     44.381M in   5.031975s
>
> Comparison:
>                 OR: :  8833775.5 i/s
>              Array: :  4920635.0 i/s - 1.80x  slower
>               Hash: :  3443628.5 i/s - 2.57x  slower
>
> ****************************
> Benchmarking value: {:a=>1, :b=>2}
>
> Warming up --------------------------------------
>              Array:    205.368k i/100ms
>               Hash:    193.242k i/100ms
>                 OR:    287.559k i/100ms
> Calculating -------------------------------------
>              Array:       3.712M (± 4.3%) i/s -     18.688M in   5.044509s
>               Hash:       3.443M (± 4.8%) i/s -     17.199M in   5.008349s
>                 OR:       7.544M (± 4.0%) i/s -     37.670M in   5.001988s
>
> Comparison:
>                 OR: :  7544189.8 i/s
>              Array: :  3711993.7 i/s - 2.03x  slower
>               Hash: :  3442707.2 i/s - 2.19x  slower
>
> After launching it several times I always get similar results. As there
> are only 3 supported classes, I think I will go for the OR which always
> performs better.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/33465#discussion_r207735370>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AADpYGds0tX_47jWJUupBtEsyrVXh8btks5uNupdgaJpZM4VlNiR>
> .
>

schneems commented:
Sorry I meant compare by identity them rehash
https://ruby-doc.org/core-2.5.1/Hash.html#method-i-compare_by_identity
On Sun, Aug 5, 2018 at 9:57 AM richard schneeman <
richard.schneeman@gmail.com> wrote:

> Call rehash on the hash before running the tests also try adding a 4th key
> since we plan on adding an “escape valve” class.
>
>
> On Sun, Aug 5, 2018 at 7:51 AM Alberto Almagro Sotelo <
> notifications@github.com> wrote:
>
>> *@albertoalmagro* commented on this pull request.
>> ------------------------------
>>
>> In actionpack/lib/action_dispatch/middleware/flash.rb
>> <https://github.com/rails/rails/pull/33465#discussion_r207735370>:
>>
>> >          def stringify_array(array) # :doc:
>>            array.map do |item|
>>              item.kind_of?(Symbol) ? item.to_s : item
>>            end
>>          end
>> +
>> +        def supported_primitive_type?(v)
>> +          v.nil? || SUPPORTED_PRIMITIVE_TYPES.any? { |type| v.is_a?(type) }
>>
>> OK, I got it wrong, this should give us a hint, now using benchmark-ips
>> gem ❤️
>>
>> begin
>>   require "bundler/inline"rescue LoadError => e
>>   $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
>>   raise eend
>>
>> gemfile(true) do
>>   source "https://rubygems.org"
>>
>>   git_source(:github) { |repo| "https://github.com/#{repo}.git" }
>>
>>   gem "benchmark-ips"end
>> require "benchmark/ips"
>> SUPPORTED_PRIMITIVE_ARRAY = [String, Array, Hash].freezeSUPPORTED_PRIMITIVE_HASH  = { String => true, Array => true, Hash => true }
>>
>> ["String", [1, 2, 3], { a: 1, b: 2}].each do |value|
>>   puts "\n****************************"
>>   puts "Benchmarking value: #{value}"
>>   puts
>>
>>   Benchmark.ips do |test|
>>     test.report("Array: ") do
>>       SUPPORTED_PRIMITIVE_ARRAY.any? { |type| value.is_a?(type) }
>>     end
>>
>>     test.report("Hash: ") do
>>       SUPPORTED_PRIMITIVE_HASH[value.class]
>>     end
>>
>>     test.report("OR: ") do
>>       value.is_a?(String) || value.is_a?(Array) || value.is_a?(Hash)
>>     end
>>
>>     test.compare!
>>   endend
>>
>> And the results:
>>
>> 2.5.1 ~ $ ruby bench_array_hash_or.rb
>> Fetching gem metadata from https://rubygems.org/.
>> Resolving dependencies...
>> Using benchmark-ips 2.7.2
>> Using bundler 1.16.1
>> ****************************
>> Benchmarking value: String
>>
>> Warming up --------------------------------------
>>              Array:    272.789k i/100ms
>>               Hash:    192.142k i/100ms
>>                 OR:    336.287k i/100ms
>> Calculating -------------------------------------
>>              Array:       6.625M (± 4.6%) i/s -     33.280M in   5.035105s
>>               Hash:       3.442M (± 4.2%) i/s -     17.293M in   5.033748s
>>                 OR:      10.453M (± 7.6%) i/s -     52.124M in   5.024647s
>>
>> Comparison:
>>                 OR: : 10453063.7 i/s
>>              Array: :  6624951.4 i/s - 1.58x  slower
>>               Hash: :  3441698.7 i/s - 3.04x  slower
>>
>> ****************************
>> Benchmarking value: [1, 2, 3]
>>
>> Warming up --------------------------------------
>>              Array:    214.251k i/100ms
>>               Hash:    183.673k i/100ms
>>                 OR:    295.872k i/100ms
>> Calculating -------------------------------------
>>              Array:       4.921M (± 4.7%) i/s -     24.639M in   5.018891s
>>               Hash:       3.444M (± 3.8%) i/s -     17.265M in   5.021349s
>>                 OR:       8.834M (± 3.8%) i/s -     44.381M in   5.031975s
>>
>> Comparison:
>>                 OR: :  8833775.5 i/s
>>              Array: :  4920635.0 i/s - 1.80x  slower
>>               Hash: :  3443628.5 i/s - 2.57x  slower
>>
>> ****************************
>> Benchmarking value: {:a=>1, :b=>2}
>>
>> Warming up --------------------------------------
>>              Array:    205.368k i/100ms
>>               Hash:    193.242k i/100ms
>>                 OR:    287.559k i/100ms
>> Calculating -------------------------------------
>>              Array:       3.712M (± 4.3%) i/s -     18.688M in   5.044509s
>>               Hash:       3.443M (± 4.8%) i/s -     17.199M in   5.008349s
>>                 OR:       7.544M (± 4.0%) i/s -     37.670M in   5.001988s
>>
>> Comparison:
>>                 OR: :  7544189.8 i/s
>>              Array: :  3711993.7 i/s - 2.03x  slower
>>               Hash: :  3442707.2 i/s - 2.19x  slower
>>
>> After launching it several times I always get similar results. As there
>> are only 3 supported classes, I think I will go for the OR which always
>> performs better.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/rails/rails/pull/33465#discussion_r207735370>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/AADpYGds0tX_47jWJUupBtEsyrVXh8btks5uNupdgaJpZM4VlNiR>
>> .
>>
>

albertoalmagro commented:
@schneems I'm sorry for your screen, you are doing it great with response time, don't worry about that. I have been thinking on a configuration variable instead of a class as "escape valve" mechanism. Don't you think it would it be less painful to set the variable while they get this fixed than packing everything into a class? With a class we would also have to take into account how adding this extra class affects serialization/deserialization.

albertoalmagro commented:
@schneems Thanks a lot for the performance lesson about Hash 📖  It is indeed much faster 👏 

I have substituted the deprecation warnings, now they are raising an `ArgumentError` when another type is provided. I hope this was what you meant.

On the last topic to address here: What do you think about using the configuration "escape valve" instead of the wrapper class as I suggested here? https://github.com/rails/rails/pull/33465#issuecomment-410527840

albertoalmagro commented:
@schneems Another topic regarding test error at `railties/test/application/assets_test.rb:376`. What about supporting trivial cases like Symbol, Numeric and Boolean already in this PR?

rafaelfranca commented:
> About the deprecation: we didn’t do a great job of letting you know what we wanted. In this case rails master is pointing at version 6.0. What we want is for the behavior in master to raise an error.
> 
> Then in a separate PR we can add a deprecation notice and target the 5.2 branch. So that anyone upgrading will first get the deprecation notice before they upgrade and start getting errors.

We can't add deprecation to the 5.2 release anymore. We can only deprecate on 6.0 and remove on 6.1

albertoalmagro commented:
@rafaelfranca @schneems I had a deprecation note before pointing that this would be removed on 6.1

Do you prefer to raise the error or shall I put back the deprecation note again?

rafaelfranca commented:
We should have the deprecation pointing to 6.1

albertoalmagro commented:
I would then suggest adding the depreciation note pointing to 6.1, which would make unnecessary adding a “escape valve” mechanism (as we would warn during 6.0 version) giving enough tyne to adapt. In 6.1 we would then remove the depreciation note and raise.

I would also suggest keeping the extra allowed types I’ve added now.

@rafaelfranca @schneems please tell me if you agree with this.

schneems commented:
The extra types are a good idea as long as they serialize fine.

We need the escape valve now so people who want to run without deprecations (and use that feature) can. We should also freeze the hash constant and make it private https://aaronlasseigne.com/2016/10/26/know-ruby-private_constant/. 

For a configuration based escape valve I’m okay with it pending implementation.

Thanks for hanging in there. This is an iterative process and we are moving forwards. Think of it as ultra slow async pair programming 😉

albertoalmagro commented:
**Done:**
* Freeze constant and add private_constant (thanks @schneems) ❤️ (squashed into d753c5c)
* Added deprecation warning for Rails 6.1.

**Still TODO:**
* Add escape valve to the configuration to skip deprecation warning.
* I also want to add some tests for additional supported types serialization.

@schneems I have my doubts on the escape valve when it comes together with a deprecation warning. What I am thinking about is that if we deprecate something, is because we don't want anyone to use it in the future and therefore we warn them about it over a period of time. I think we shouldn't allow them to hide the deprecation warning. If we do, users could forget about it and on Rails 6.1, when storing custom objects finally gets deprecated, an error would be raised.

matthewd commented:
I don't follow why we'd add this warning specifically on flash, rather than applying it to everything stored in the session (and thus pick up flash indirectly). It seems like a layering violation for flash to be enforcing storage rules because of technical limitations of something lower in the stack. Am I missing something?

albertoalmagro commented:
Hi @matthewd ! Good point, thanks for mentioning this. We are only focusing on the flash because the related issue came from that side and Flash's documentation specified that only primitive types like String, Array and Hash could be stored there (see docs [here](https://github.com/rails/rails/blob/master/actionpack/lib/action_dispatch/middleware/flash.rb#L6)).

While I didn't find a similar hint on Session's documentation, now that you mention that I think that the same restrictions should also apply to it as the flash itself is stored in the session. What do you think @rafaelfranca @schneems ?

kaspth commented:
I think @matthewd is also alluding to https://github.com/rails/activerecord-session_store, which doesn't have the type restrictions that the default session store has.

The possible maintenance burden on this outweighs the benefits for me. But that's on a skim.

schneems commented:
From a perf perspective Sprockets has a somewhat similar type check and it’s pretty expensive (though it may get called many times per asset). I was thinking his would be fine if it’s only triggered once per request and only if flash is being used.

On that note it could perhaps be a setting that is enabled by default on development and disabled in prod (to mitigate any possible performance issue).

Regarding maintenance burden my preference is generally to guard behavior that has bad results with difficult to debug behavior. The check itself isn’t really that complicated (IMHO). If we make the escape valve a global config then people can also turn it off if they’re using active record session store. Though the question then becomes if we are overloaded on config values.

My preference would be to move forward here and make it a general check that is on in Test and dev by default and off in production configurable to be disabled. I’ll defer to MD and Kasper though if you all want to give it another look and let me know what you think.

Alberto: I know it’s frystrating to work for awhile on a feature that might not make it in, which might be the case here. However, it’s not for nothing: you’ve got some good experience in the Rails codebase that will help you out with any future patches and you’ve gotten some perf experience as well! No matter which way this thing goes, thanks for the PR and thanks for all your work here!

matthewd commented:
I can see @kaspth's concern about maintenance... and while accurate docs are accurate, I don't like the aesthetics of explicitly listing all the allowed classes (TrueClass & friends, especially).

But I can also see the UX benefit to not accepting a value we won't later accurately reproduce. I'd like to consider a version of this that lives at the session (or cookie?) level... if we can buy more coverage (both in consistency and in number of places a user could do the surprising thing) for a comparable amount of code, I think that shifts the cost/benefit balance.

albertoalmagro commented:
Hi everyone! Sorry for the delay, I'm back from holidays and EuRuKo 🎉 

@schneems I've just pushed some work I had almost finished before leaving regarding the escape valve mechanism f717b50

@schneems @matthewd @kaspth I see the point in your comments. If you prefer to have this check for the whole session values (only when CookieStore is used) I don't mind throwing away what I've already done here for the flash, I prefer always to have a better framework and I know you as core team always give the best advice to achieve it. Just tell me 😉 I would however integrate this commit 75dfb0c9e47f1207bcb91b0fdd0e44a34f81b5d2 as I think it helps avoiding the issue that started this in the meanwhile.

albertoalmagro commented:
@schneems @matthewd @kaspth I leave this PR as it originally was, I will address the CookieStore/Session type check issue in another PR. Thanks!

oniofchaos commented:
Good idea to document 👍 

Do you know if there is a test that already memorializes this behavior? If not, might be a good idea to add that as well.

albertoalmagro commented:
@oniofchaos nope, there isn't a test at the moment. As it was commented in this ticket, I am going to develop it for the whole session but in another PR.

@schneems does this need something else? Thanks!

albertoalmagro commented:
Review please 🙏
